### ISSUE TITLE
CA1810: false positive when field is used in static ctor

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn-analyzers#3852]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.3.0-beta1.final (Latest pre-release)

#### Diagnostic ID

Example: [CA1810](https://docs.microsoft.com/visualstudio/code-quality/ca1810)

#### Repro steps

```csharp
[Fact]
public async Task CA1810_FieldMethodCall()
{
	await new VerifyCS.Test
	{
		ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithWinForms,
		TestCode = @"
using System;
using System.Windows.Forms;

public class SKWindow
{
    public void CreateControl() {}
}

public class SendKeys
{
    private static readonly SKWindow messageWindow;

    static SendKeys()
    {
        Application.ThreadExit += new EventHandler(OnThreadExit);
        messageWindow = new SKWindow();
        messageWindow.CreateControl();
    }

    private static void OnThreadExit(object sender, EventArgs e) {}
}
",
	}.RunAsync();
}
```

#### Expected behavior

No CA1810 diagnostic.

#### Actual behavior

CA1810 diagnostic.

See https://github.com/dotnet/winforms/pull/3561/files

**Comment by Evangelink** [comment]
Looking at the issue, I think that there are a couple of cases/questions raised here.

## Scenario 1

```csharp
public class C
{
	private static readonly SKWindow messageWindow;

	static C()
	{
		messageWindow = new SKWindow();
        messageWindow.CreateControl();
	}
}
```

In this scenario, I think that an issue shall be reported because nothing prevents you from creating a static method which returns the object initialized. For example a solution could be:

```csharp
public class C
{
	private static readonly SKWindow messageWindow = GetMessageWindow();

	private static SKWindow GetMessageWindow()
	{
		var window = new SKWindow();
        window.CreateControl();
		return window;
	}
}
```

## Scenario 2

This scenario is IMO the actual problem here. As long as you have a subscription (`+=` or `-=`) for an event (what about actions?) in the ctor, don't report no matter what. Anyways you need to have the static ctor.

## Scenario 3

Initialization order matters, let's say you have a couple of fields and you need to chain some calls on these fields in a particular order. In this case, I don't think there is anything that can be one automatically so users will have to suppress the diagnostic.

```csharp
public class C
{
	private static readonly A fieldA;
	private static readonly B fieldB;

	static C()
	{
		fieldA = new A();
		fieldA.DoSomething();

		fieldB = new B();
		fieldB.DoSomething();

		fieldA.DoSomethingElseWhichDependsUpon(fieldB);
	}
}
```
